[Build 113](https://ci.wildfly.org/viewLog.html?buildId=170173&buildTypeId=Xnio_33_LinuxIbmJdk8) outcome was **FAILURE** using a merge of b176feb8e7866b75cc8a90897d0593f0e6f06e26
Summary: Tests failed: 7 (1 new), passed: 950, ignored: 18 Build time: 00:01:11
### Failed tests
```
org.xnio.nio.test.MultiThreadedNioSslTcpConnectionTestCase.twoWayTransfer: java.lang.AssertionError
	at org.xnio.nio.test.AbstractNioSslTcpTest$7.run(AbstractNioSslTcpTest.java:240)
	at org.xnio.nio.test.AbstractNioTcpTest.doConnectionTest(AbstractNioTcpTest.java:102)
	at org.xnio.nio.test.NioSslTcpConnectionTestCase.doConnectionTest(NioSslTcpConnectionTestCase.java:115)
	at org.xnio.nio.test.AbstractNioSslTcpTest.twoWayTransfer(AbstractNioSslTcpTest.java:237)
------- Stderr: -------
20:15:02,386 INFO  (main) [TEST] <AbstractNioSslTcpTest.java:231> Test: twoWayTransfer
20:15:02,409 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:255> client handle readable
20:15:02,416 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,416 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,417 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,417 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,417 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,417 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,418 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,418 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,418 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,419 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,419 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,419 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,419 INFO  (XNIO-45 I/O-1) [TEST] <AbstractNioSslTcpTest.java:336> server handle readable
20:15:02,419 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,420 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,420 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,420 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,420 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,421 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,421 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,421 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,421 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,422 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,422 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,423 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,424 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,424 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,424 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,424 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,425 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,425 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,425 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,425 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,426 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,426 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,426 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,427 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,427 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,427 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,427 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,428 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,428 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable
20:15:02,428 INFO  (XNIO-45 I/O-2) [TEST] <AbstractNioSslTcpTest.java:273> client handle writable


org.xnio.racecondition.ResumeReadsOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.ResumeWritesOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetReadListenerOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetReadReadyOnHandlingReadableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetWriteListenerOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)


org.xnio.racecondition.SetWriteReadyOnHandlingWritableChannelTestCase: java.net.ConnectException: Connection refused (Connection refused)
------- Stdout: -------
byteman jar is /store/repository/org/jboss/byteman/byteman/4.0.6/byteman-4.0.6.jar
Setting org.jboss.byteman.allow.config.update=true
------- Stderr: -------
	at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:408)
	at sun.instrument.InstrumentationImpl.loadClassAndCallAgentmain(InstrumentationImpl.java:433)
	at com.ibm.tools.attach.javaSE.Attachment.loadAgentLibraryImpl(Native Method)
	at com.ibm.tools.attach.javaSE.Attachment.loadAgentLibrary(Attachment.java:309)
	at com.ibm.tools.attach.javaSE.Attachment.parseLoadAgent(Attachment.java:287)
	at com.ibm.tools.attach.javaSE.Attachment.doCommand(Attachment.java:168)
	at com.ibm.tools.attach.javaSE.Attachment.run(Attachment.java:128)
Caused by: java.lang.UnsupportedOperationException: adding retransformable transformers is not supported in this environment
	at sun.instrument.InstrumentationImpl.addTransformer(InstrumentationImpl.java:100)
	at org.jboss.byteman.agent.Main.premain(Main.java:279)
	at org.jboss.byteman.agent.Main.agentmain(Main.java:309)
	... 11 more


```

--- a/nio-impl/pom.xml
+++ b/nio-impl/pom.xml
@@ -40,6 +40,7 @@
         <xnio.nio.selector.main/>
         <xnio.nio.selector.temp/>
         <xnio.nio.selector.provider/>
+        <xnio.nio.alt-queued-server>true</xnio.nio.alt-queued-server>
     </properties>
 
     <dependencies>
@@ -154,6 +155,10 @@
                         <property>
                             <name>org.xnio.ssl.new</name>
                             <value>${org.xnio.ssl.new}</value>
+                        </property>
+                        <property>
+                            <name>xnio.nio.alt-queued-server</name>
+                            <value>${xnio.nio.alt-queued-server}</value>
                         </property>
                     </systemProperties>
                     <enableAssertions>true</enableAssertions>

--- a/nio-impl/src/main/java/org/xnio/nio/Log.java
+++ b/nio-impl/src/main/java/org/xnio/nio/Log.java
@@ -125,6 +125,10 @@ interface Log extends BasicLogger {
     @Message(id = 8000, value = "Received an I/O error on selection: %s")
     void selectionError(IOException e);
 
+    @LogMessage(level = WARN)
+    @Message(id = 8001, value = "Socket accept failed, backing off for %2$d milliseconds: %1$s")
+    void acceptFailed(IOException problem, int backOffTime);
+
     // Trace
 
     @LogMessage(level = TRACE)

--- a/nio-impl/src/main/java/org/xnio/nio/NioTcpServer.java
+++ b/nio-impl/src/main/java/org/xnio/nio/NioTcpServer.java
@@ -112,24 +112,35 @@ final class NioTcpServer extends AbstractNioChannel<NioTcpServer> implements Acc
 
     private static final AtomicLongFieldUpdater<NioTcpServer> connectionStatusUpdater = AtomicLongFieldUpdater.newUpdater(NioTcpServer.class, "connectionStatus");
 
-    NioTcpServer(final NioXnioWorker worker, final ServerSocketChannel channel, final OptionMap optionMap) throws IOException {
+    NioTcpServer(final NioXnioWorker worker, final ServerSocketChannel channel, final OptionMap optionMap, final boolean useAcceptThreadOnly) throws IOException {
         super(worker);
         this.channel = channel;
-        final WorkerThread[] threads = worker.getAll();
-        final int threadCount = threads.length;
-        if (threadCount == 0) {
-            throw log.noThreads();
-        }
-        final int tokens = optionMap.get(Options.BALANCING_TOKENS, -1);
-        final int connections = optionMap.get(Options.BALANCING_CONNECTIONS, 16);
-        if (tokens != -1) {
-            if (tokens < 1 || tokens >= threadCount) {
-                throw log.balancingTokens();
+        final WorkerThread[] threads;
+        final int threadCount;
+        final int tokens;
+        final int connections;
+        if (useAcceptThreadOnly) {
+            threads = new WorkerThread[] { worker.getAcceptThread() };
+            threadCount = 1;
+            tokens = 0;
+            connections = 0;
+        } else {
+            threads = worker.getAll();
+            threadCount = threads.length;
+            if (threadCount == 0) {
+                throw log.noThreads();
             }
-            if (connections < 1) {
-                throw log.balancingConnectionCount();
+            tokens = optionMap.get(Options.BALANCING_TOKENS, -1);
+            connections = optionMap.get(Options.BALANCING_CONNECTIONS, 16);
+            if (tokens != -1) {
+                if (tokens < 1 || tokens >= threadCount) {
+                    throw log.balancingTokens();
+                }
+                if (connections < 1) {
+                    throw log.balancingConnectionCount();
+                }
+                tokenConnectionCount = connections;
             }
-            tokenConnectionCount = connections;
         }
         socket = channel.socket();
         if (optionMap.contains(Options.SEND_BUFFER)) {
@@ -383,7 +394,12 @@ final class NioTcpServer extends AbstractNioChannel<NioTcpServer> implements Acc
         if (current == null) {
             return null;
         }
-        final NioTcpServerHandle handle = handles[current.getNumber()];
+        final NioTcpServerHandle handle;
+        if (handles.length == 1) {
+            handle = handles[0];
+        } else {
+            handle = handles[current.getNumber()];
+        }
         if (! handle.getConnection()) {
             return null;
         }

--- a/nio-impl/src/main/java/org/xnio/nio/NioTcpServer.java
+++ b/nio-impl/src/main/java/org/xnio/nio/NioTcpServer.java
@@ -424,6 +424,7 @@ final class NioTcpServer extends AbstractNioChannel<NioTcpServer> implements Acc
                 newConnection.setOption(Options.READ_TIMEOUT, Integer.valueOf(readTimeout));
                 newConnection.setOption(Options.WRITE_TIMEOUT, Integer.valueOf(writeTimeout));
                 ok = true;
+                handle.resetBackOff();
                 return newConnection;
             } finally {
                 if (! ok) safeClose(accepted);
@@ -431,6 +432,10 @@ final class NioTcpServer extends AbstractNioChannel<NioTcpServer> implements Acc
         } catch (ClosedChannelException e) {
             throw e;
         } catch (IOException e) {
+            // something went wrong with the accept
+            // it could be due to running out of file descriptors, or due to closed channel, or other things
+            handle.startBackOff();
+            log.acceptFailed(e, handle.getBackOffTime());
             return null;
         } finally {
             if (! ok) {

--- a/nio-impl/src/main/java/org/xnio/nio/NioTcpServerHandle.java
+++ b/nio-impl/src/main/java/org/xnio/nio/NioTcpServerHandle.java
@@ -19,10 +19,14 @@
 package org.xnio.nio;
 
 import java.nio.channels.SelectionKey;
+import java.util.concurrent.TimeUnit;
+
 import org.xnio.ChannelListeners;
 
+import static java.lang.Math.min;
 import static java.lang.Thread.currentThread;
 import static org.xnio.IoUtils.safeClose;
+import static sun.swing.MenuItemLayoutHelper.max;
 
 /**
 * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
@@ -36,6 +40,8 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
     private int high;
     private int tokenCount = -1;
     private boolean stopped;
+    private boolean backOff;
+    private int backOffTime = 0;
 
     NioTcpServerHandle(final NioTcpServer server, final SelectionKey key, final WorkerThread thread, final int low, final int high) {
         super(thread, key);
@@ -68,7 +74,7 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
     void resume() {
         final WorkerThread thread = getWorkerThread();
         if (thread == currentThread()) {
-            if (! stopped && server.resumed) super.resume(SelectionKey.OP_ACCEPT);
+            if (! stopped && ! backOff && server.resumed) super.resume(SelectionKey.OP_ACCEPT);
         } else {
             thread.execute(new Runnable() {
                 public void run() {
@@ -81,7 +87,7 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
     void suspend() {
         final WorkerThread thread = getWorkerThread();
         if (thread == currentThread()) {
-            if (stopped || ! server.resumed) super.suspend(SelectionKey.OP_ACCEPT);
+            if (stopped || backOff || ! server.resumed) super.suspend(SelectionKey.OP_ACCEPT);
         } else {
             thread.execute(new Runnable() {
                 public void run() {
@@ -105,6 +111,8 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
         if (count-- <= low && tokenCount != 0 && stopped) {
             stopped = false;
             if (server.resumed) {
+                // end backoff optimistically
+                backOff = false;
                 super.resume(SelectionKey.OP_ACCEPT);
             }
         }
@@ -117,7 +125,7 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
                 tokenCount = newCount;
                 if (count <= low && stopped) {
                     stopped = false;
-                    if (server.resumed) {
+                    if (server.resumed && ! backOff) {
                         super.resume(SelectionKey.OP_ACCEPT);
                     }
                 }
@@ -126,6 +134,31 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
             workerThread = workerThread.getNextThread();
         }
         setThreadNewCount(workerThread, newCount);
+    }
+
+    /**
+     * Start back-off, when an accept produces an exception.
+     */
+    void startBackOff() {
+        backOff = true;
+        backOffTime = max(250, min(30_000, backOffTime << 2));
+        suspend();
+        getWorkerThread().executeAfter(this::endBackOff, backOffTime, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * End back-off, when an accept may be retried.
+     */
+    void endBackOff() {
+        backOff = false;
+        resume();
+    }
+
+    /**
+     * Reset back-off, when an accept has succeeded.
+     */
+    void resetBackOff() {
+        backOffTime = 0;
     }
 
     private void setThreadNewCount(final WorkerThread workerThread, final int newCount) {
@@ -157,7 +190,7 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
 
     boolean getConnection() {
         assert currentThread() == getWorkerThread();
-        if (stopped) {
+        if (stopped || backOff) {
             return false;
         }
         if (tokenCount != -1 && --tokenCount == 0) {
@@ -180,7 +213,7 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
                 suspend();
             } else if (count <= low && stopped) {
                 stopped = false;
-                if (server.resumed) resume();
+                if (server.resumed && ! backOff) resume();
             }
         } else {
             thread.execute(new Runnable() {
@@ -194,5 +227,9 @@ final class NioTcpServerHandle extends NioHandle implements ChannelClosed {
     int getConnectionCount() {
         assert currentThread() == getWorkerThread();
         return count;
+    }
+
+    int getBackOffTime() {
+        return backOffTime;
     }
 }

--- a/nio-impl/src/main/java/org/xnio/nio/NioXnio.java
+++ b/nio-impl/src/main/java/org/xnio/nio/NioXnio.java
@@ -47,6 +47,7 @@ final class NioXnio extends Xnio {
 
     static final boolean IS_HP_UX;
     static final boolean HAS_BUGGY_EVENT_PORT;
+    static final boolean USE_ALT_QUEUED_SERVER;
 
     interface SelectorCreator {
         Selector open() throws IOException;
@@ -64,6 +65,7 @@ final class NioXnio extends Xnio {
                 return Boolean.valueOf(System.getProperty("os.name", "unknown").equalsIgnoreCase("hp-ux"));
             }
         }).booleanValue();
+        USE_ALT_QUEUED_SERVER = Boolean.parseBoolean(AccessController.doPrivileged(new ReadPropertyAction("xnio.nio.alt-queued-server", "true")));
         // if a JDK is released with a fix, we can try to detect it and set this to "false" for those JDKs.
         HAS_BUGGY_EVENT_PORT = true;
     }

--- a/nio-impl/src/main/java/org/xnio/nio/NioXnioWorker.java
+++ b/nio-impl/src/main/java/org/xnio/nio/NioXnioWorker.java
@@ -180,7 +180,12 @@ final class NioXnioWorker extends XnioWorker {
                 channel.socket().bind(bindAddress);
             }
             if (false) {
-                final NioTcpServer server = new NioTcpServer(this, channel, optionMap);
+                final NioTcpServer server = new NioTcpServer(this, channel, optionMap, false);
+                server.setAcceptListener(acceptListener);
+                ok = true;
+                return server;
+            } else if (NioXnio.USE_ALT_QUEUED_SERVER) {
+                final QueuedNioTcpServer2 server = new QueuedNioTcpServer2(new NioTcpServer(this, channel, optionMap, true));
                 server.setAcceptListener(acceptListener);
                 ok = true;
                 return server;

--- a/nio-impl/src/main/java/org/xnio/nio/QueuedNioTcpServer2.java
+++ b/nio-impl/src/main/java/org/xnio/nio/QueuedNioTcpServer2.java
@@ -1,0 +1,174 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2019 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.xnio.nio;
+
+import java.io.IOException;
+import java.net.SocketAddress;
+import java.nio.channels.ClosedChannelException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.wildfly.common.Assert;
+import org.xnio.ChannelListener;
+import org.xnio.ChannelListeners;
+import org.xnio.Option;
+import org.xnio.StreamConnection;
+import org.xnio.XnioExecutor;
+import org.xnio.XnioIoThread;
+import org.xnio.channels.AcceptListenerSettable;
+import org.xnio.channels.AcceptingChannel;
+
+final class QueuedNioTcpServer2 extends AbstractNioChannel<QueuedNioTcpServer2> implements AcceptingChannel<StreamConnection>, AcceptListenerSettable<QueuedNioTcpServer2> {
+    private final NioTcpServer realServer;
+    private final List<Queue<StreamConnection>> acceptQueues;
+
+    private final Runnable acceptTask = this::acceptTask;
+
+    private volatile ChannelListener<? super QueuedNioTcpServer2> acceptListener;
+
+    QueuedNioTcpServer2(final NioTcpServer realServer) {
+        super(realServer.getWorker());
+        this.realServer = realServer;
+        final NioXnioWorker worker = realServer.getWorker();
+        final int cnt = worker.getIoThreadCount();
+        acceptQueues = new ArrayList<>(cnt);
+        for (int i = 0; i < cnt; i ++) {
+            acceptQueues.add(new LinkedBlockingQueue<>());
+        }
+        realServer.getCloseSetter().set(ignored -> invokeCloseHandler());
+        realServer.getAcceptSetter().set(ignored -> handleReady());
+    }
+
+    public StreamConnection accept() throws IOException {
+        final WorkerThread current = WorkerThread.getCurrent();
+        if (current == null) {
+            return null;
+        }
+        final Queue<StreamConnection> socketChannels = acceptQueues.get(current.getNumber());
+        final StreamConnection connection = socketChannels.poll();
+        if (connection == null) {
+            if (! realServer.isOpen()) {
+                throw new ClosedChannelException();
+            }
+        }
+        return connection;
+    }
+
+    public ChannelListener<? super QueuedNioTcpServer2> getAcceptListener() {
+        return acceptListener;
+    }
+
+    public void setAcceptListener(final ChannelListener<? super QueuedNioTcpServer2> listener) {
+        this.acceptListener = listener;
+    }
+
+    public ChannelListener.Setter<QueuedNioTcpServer2> getAcceptSetter() {
+        return new Setter<QueuedNioTcpServer2>(this);
+    }
+
+    public SocketAddress getLocalAddress() {
+        return realServer.getLocalAddress();
+    }
+
+    public <A extends SocketAddress> A getLocalAddress(final Class<A> type) {
+        return realServer.getLocalAddress(type);
+    }
+
+    public void suspendAccepts() {
+        realServer.suspendAccepts();
+    }
+
+    public void resumeAccepts() {
+        realServer.resumeAccepts();
+    }
+
+    public boolean isAcceptResumed() {
+        return realServer.isAcceptResumed();
+    }
+
+    public void wakeupAccepts() {
+        realServer.wakeupAccepts();
+    }
+
+    public void awaitAcceptable() {
+        throw Assert.unsupported();
+    }
+
+    public void awaitAcceptable(final long time, final TimeUnit timeUnit) {
+        throw Assert.unsupported();
+    }
+
+    @Deprecated
+    public XnioExecutor getAcceptThread() {
+        return getIoThread();
+    }
+
+    public void close() throws IOException {
+        realServer.close();
+    }
+
+    public boolean isOpen() {
+        return realServer.isOpen();
+    }
+
+    public boolean supportsOption(final Option<?> option) {
+        return realServer.supportsOption(option);
+    }
+
+    public <T> T getOption(final Option<T> option) throws IOException {
+        return realServer.getOption(option);
+    }
+
+    public <T> T setOption(final Option<T> option, final T value) throws IllegalArgumentException, IOException {
+        return realServer.setOption(option, value);
+    }
+
+    void handleReady() {
+        NioSocketStreamConnection connection;
+        try {
+            connection = realServer.accept();
+        } catch (ClosedChannelException e) {
+            return;
+        }
+        while (connection != null) {
+            final XnioIoThread thread = connection.getIoThread();
+            final Queue<StreamConnection> queue = acceptQueues.get(thread.getNumber());
+            queue.add(connection);
+            thread.execute(acceptTask);
+            try {
+                connection = realServer.accept();
+            } catch (ClosedChannelException e) {
+                return;
+            }
+        }
+    }
+
+    void acceptTask() {
+        final WorkerThread current = WorkerThread.getCurrent();
+        assert current != null;
+        final Queue<StreamConnection> queue = acceptQueues.get(current.getNumber());
+        ChannelListeners.invokeChannelListener(QueuedNioTcpServer2.this, getAcceptListener());
+        if (! queue.isEmpty()) {
+            current.execute(acceptTask);
+        }
+    }
+}

